roq-core/src/main/java/org/roqmessaging/core/factory/RoQConnectionFactory.java

```diff
@@ -102,17 +144,63 @@ private void closeSocketConnection() {
 		this.logger.debug("Closing factory socket");
 		this.globalConfigReq.setLinger(0);
 		this.globalConfigReq.close();
-		
 	}
 
 	/**
 	 * Initialise the socket connection.
 	 */
-	private void initSocketConnection() {
+	private void initSocketConnection() 
+			throws IllegalStateException {
+		// Get the active master address
+		this.configServer = zk.getGCMLeaderAddress();
+		if (configServer == null) 
+			throw new IllegalStateException("GCM node not found");
+		logger.info("Active GCM address: " + this.configServer);
 		context = ZMQ.context(1);
 		globalConfigReq = context.socket(ZMQ.REQ);
 		globalConfigReq.connect("tcp://"+this.configServer+":5000");
-		
+		globalConfigReq.setReceiveTimeOut(timeout);
+	}
+
+	/**
+	 * This method sends a request to the GCM.
+	 * First it get the active GCM, then it sent the request
+	 * If the request fails because leader has changed or the connection
+	 * cannot be established (timeout), the request is resent up to maxRetry times.
+	 * @param request
+	 * @return response
+	 * @throws ConnectException 
+	 */
+	private byte[]  sendRequest (byte[] request) throws ConnectException {
+		// The configuration should load the information about the monitor corresponding to this queue
+		byte[] response = null;
+		String responseString = "";
+		int retry = 0;
+		// If the request has failed, we retry until to reach the maxRetry
+		do {
+			try {
+				if (retry > 0) {
+					logger.info("GCM not found");
+					Thread.sleep(3000); // Wait between two requests
```

Comment:
You already have the timeout which ensures you don't spam requests. Why wait three additional seconds? Is it necessary?